internal/encoding/asn1/asn1.go

```diff
@@ -0,0 +1,219 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package asn1 decodes BER-encoded ASN.1 data structures and encodes in DER.
+// Note: DER is a subset of BER.
+// Reference: http://luca.ntop.org/Teaching/Appunti/asn1.html
+package asn1
+
+import (
+	"bytes"
+	"encoding/asn1"
+	"math"
+)
+
+// Common errors
+var (
+	ErrEarlyEOF                   = asn1.SyntaxError{Msg: "early EOF"}
+	ErrTrailingData               = asn1.SyntaxError{Msg: "trailing data"}
+	ErrUnsupportedLength          = asn1.StructuralError{Msg: "length method not supported"}
+	ErrUnsupportedIndefinedLength = asn1.StructuralError{Msg: "indefinite length not supported"}
+)
+
+// value represents an ASN.1 value.
+type value interface {
+	// Encode encodes the value to the value writer in DER.
+	Encode(*bytes.Buffer) error
+
+	// EncodedLen returns the length in bytes of the encoded data.
+	EncodedLen() int
+}
+
+// ConvertToDER converts BER-encoded ASN.1 data structures to DER-encoded.
+func ConvertToDER(ber []byte) ([]byte, error) {
+	v, err := decode(ber)
+	if err != nil {
+		return nil, err
+	}
+	buf := bytes.NewBuffer(make([]byte, 0, v.EncodedLen()))
+	if err = v.Encode(buf); err != nil {
```

Comment:
Encoding is still in a recursive manner. Can we make it iterative?

Reply:
Updated to remove recusive function call.